### PR TITLE
wasm: Add WAMR support

### PR DESCRIPTION
**Pull request by leyao-daily**
Commit Message: Enabling WAMR(Webassembly Micro Runtime) as wasm runtime in Envoy
Additional Description:
```
WebAssembly Micro Runtime (WAMR) is a standalone WebAssembly (WASM) runtime with a small footprint. It includes a few parts as below:

The "iwasm" VM core, supporting WebAssembly interpreter, ahead of time compilation (AoT) and Just-in-Time compilation (JIT)

The application framework and the supporting API's for the WASM applications

The dynamic management of the WASM applications
```
Risk Level: Medium 
Testing: Runtime unit testing, integration testing and manual testing
```
...
//test/integration:http2_flood_integration_test                          PASSED in 13.5s
  Stats over 4 runs: max = 13.5s, min = 11.3s, avg = 12.7s, dev = 0.8s
//test/integration:multiplexed_integration_test                          PASSED in 19.1s
  Stats over 4 runs: max = 19.1s, min = 13.5s, avg = 16.5s, dev = 2.0s
//test/common/http/http2:codec_impl_test                                 PASSED in 4.9s
  Stats over 5 runs: max = 4.9s, min = 3.6s, avg = 4.2s, dev = 0.5s
//test/integration:protocol_integration_test                             PASSED in 40.3s
  Stats over 5 runs: max = 40.3s, min = 32.0s, avg = 35.4s, dev = 2.9s
//test/integration:quic_http_integration_test                            PASSED in 13.0s
  Stats over 6 runs: max = 13.0s, min = 9.6s, avg = 10.7s, dev = 1.2s
//test/integration:quic_protocol_integration_test                        PASSED in 40.3s
  Stats over 8 runs: max = 40.3s, min = 23.5s, avg = 29.4s, dev = 5.4s
//test/extensions/filters/http/wasm:config_test                          PASSED in 3.2s
  Stats over 50 runs: max = 3.2s, min = 0.4s, avg = 0.7s, dev = 0.6s
//test/extensions/filters/http/wasm:wasm_filter_test                     PASSED in 6.2s
  Stats over 50 runs: max = 6.2s, min = 0.9s, avg = 3.5s, dev = 1.2s

Executed 855 out of 855 tests: 855 tests pass.
```
Docs Changes: N/A
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: bazel/external/wamr.BUILD
Comment (lum1n0us) [inline bazel/external/wamr.BUILD:16]: need to remove

File: bazel/repository_locations.bzl
Comment (lum1n0us) [inline bazel/repository_locations.bzl:711]: to 0.1-beta

File: bazel/repository_locations.bzl
Comment (asraa) [inline bazel/repository_locations.bzl:714]: Why not https://github.com/bytecodealliance/wasm-micro-runtime?

Comment (leyao-daily) [reply bazel/repository_locations.bzl:714]: The changes will be merged into https://github.com/bytecodealliance/wasm-micro-runtime after finish the envoy checking.

**Review by PiotrSikora — CHANGES_REQUESTED** [review]
Thank you for working on this. You're still missing:
- Wasm Runtime factory for WAMR (see: [`source/extensions/wasm_runtime`](https://github.com/envoyproxy/envoy/tree/main/source/extensions/wasm_runtime)),
- tests executed on the CI (see: [`bazel.compile_time_options` in `ci/do_ci.sh`](https://github.com/envoyproxy/envoy/blob/a809844ddeda2fce1f5d3cdfe6204fafb7fd4452/ci/do_ci.sh#L327L343)).

File: test/extensions/bootstrap/wasm/wasm_speed_test.cc
Comment (leyao-daily) [reply test/extensions/bootstrap/wasm/wasm_speed_test.cc:99]: Ok. And the factory for wamr is https://github.com/envoyproxy/envoy/pull/16057/files#diff-85815a26c8f765868f9c242a1968320d41e9c5798cbe94358339f3d99d0a70c2

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:719]: New dependency additions need to adhere too: https://github.com/envoyproxy/envoy/blob/main/DEPENDENCY_POLICY.md /cc @envoyproxy/dependency-shepherds 

```
scorecard --repo=https://github.com/bytecodealliance/wasm-micro-runtime
                                                                                                                                          
RESULTS                                                                                                                                   
-------                                                                                                                                   
Repo: github.com/bytecodealliance/wasm-micro-runtime                                                                                      
Active: Pass 10                                                                                                                           
Branch-Protection: Fail 0                                                                                                                 
CI-Tests: Pass 10                                                                                                                         
CII-Best-Practices: Fail 10                                                                                                               
Code-Review: Pass 9                                                                                                                       
Contributors: Pass 10                                                                                                                     
Frozen-Deps: Fail 5                                                                                                                       
Fuzzing: Fail 10                                                                                                                          
Packaging: Fail 10                                                                                                                        
Pull-Requests: Pass 10                                                                                                                    
SAST: Fail 10                                                                                                                             
Security-Policy: Pass 10                                                                                                                  
Signed-Releases: Fail 0                                                                                                                   
Signed-Tags: Fail 10                                                                                                                      
```

File: bazel/repository_locations.bzl
Comment (htuch) [inline bazel/repository_locations.bzl:708]: @PiotrSikora what's our end game here? Will we eventually bring every Wasm runtime into the Envoy OSS build? Do we need this or can we be opinionated?

My main worry here is the increase in build time in CI, security risk envelope, etc.

CC @envoyproxy/dependency-shepherds

**Review by htuch — CHANGES_REQUESTED** [review]

File: bazel/repository_locations.bzl
Comment (PiotrSikora) [reply bazel/repository_locations.bzl:708]: Those are good questions.

From Proxy-Wasm C++ Host's perspective, supporting any decent Wasm runtime is desireable, since they all have a bit different features, and most implement variants of [Wasm C/C++ API](https://github.com/WebAssembly/wasm-c-api), so both implementation and maintanance have relatively small cost.

From Envoy's perspective, it's definitely not neccessary, and we could be opinionated, but since those are optional alternatives, I don't think there are too many downsides, especially since the common logic is abstracted away.

Regarding build time - Wasm runtimes are updated once a month or once a quarter, and the alternative ones are built only in `bazel.compile_time_options` target, so their artifacts should be pretty much always cached, and the overhead shouldn't be much of an issue.

Regarding security risk - we could officially support only the default Wasm runtime (currently: V8), and be upfront that there won't be any security releases for other runtimes.

Regarding WAMR specifically, it has interpreted mode, and claims very small binary size overhead and low memory usage, so it might be more desirable in Envoy Mobile than V8, for example.

Comment (htuch) [reply bazel/repository_locations.bzl:708]: OK, can we add to this PR a change to threat model https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/threat_model#core-and-extensions stating this explicitly?

I think this is a pretty important thing to get clear as we add more runtimes; we won't support anything other than V8 for now from a security perspective.

File: source/extensions/wasm_runtime/wamr/BUILD
Comment (htuch) [inline source/extensions/wasm_runtime/wamr/BUILD:16]: Can we add a comment that "This may never change from `unknown` until the threat model at https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/threat_model#core-and-extensions is updated to capture additional Wasm runtimes". This should also be done for all other Wasm runtimes other than V8.

Comment (leyao-daily) [reply source/extensions/wasm_runtime/wamr/BUILD:16]: So add an annotation above the `unknown` line to explain the meaning?

File: bazel/repository_locations.bzl
Comment (leyao-daily) [reply bazel/repository_locations.bzl]: Fix to WAMR official Repo

**Comment by leyao-daily** [comment]
The PR is waiting for https://github.com/envoyproxy/envoy/pull/16112 merging.

**Comment by PiotrSikora** [comment]
> The PR is waiting for #16112 merging.

The fix for removed `getCustomSection()` was merged in https://github.com/envoyproxy/envoy/pull/16162, so you should be able to unblock this PR when you update your PR in Proxy-Wasm C++ Host repo, and merge with `main` branch here.

File: source/extensions/wasm_runtime/wamr/BUILD
Comment (PiotrSikora) [reply source/extensions/wasm_runtime/wamr/BUILD:16]: Yes. Could you add that comment here and to Wasmtime and WAVM?

File: bazel/foreign_cc/BUILD
Comment (PiotrSikora) [reply bazel/foreign_cc/BUILD:293]: Also, what's the source of this new order? It doesn't match the output from:
```
docker run -it --rm envoyproxy/envoy-build-ubuntu:3d0491e2034287959a292806e3891fd0b7dd2703 /bin/bash -c '/opt/llvm/bin/llvm-config --libnames'
```
I think this wasn't updated since LLVM 6.0, and somehow it didn't break, but I believe that the order should match the output from the command above.

Comment (leyao-daily) [reply bazel/foreign_cc/BUILD:293]: My teammate works for this part and he is on vacation now, may responses slowly for it.

Comment (lum1n0us) [reply bazel/foreign_cc/BUILD:293]: Yes. The reordering is necessary for WAMR. The previous one leads an undefined symbol linking error when enabling WAMR. 
We also run a test to enable WAVM and it works well with the reordering list

Comment (PiotrSikora) [reply bazel/foreign_cc/BUILD:293]: That's great, thanks for testing. It sounds like the reordering is justified.

However, 2 questions still remain unanswered:
1) What's the source of this new order?
2) Does the order produced by `llvm-config --libnames` from LLVM 10 also work?

Thanks!

**Comment by lum1n0us** [comment]
it based on “llvm 10 config” and plus manually adjustment. literally “homemade”.

________________________________

File: test/per_file_coverage.sh
Comment (PiotrSikora) [inline test/per_file_coverage.sh:59]: You need to lower this to `40.0` to pass coverage:
`(NullVM + V8) / (NullVM + V8 + WAMR + Wasmtime + WAVM)` ~= `2 / 5` ~= `40%`

**Comment by lum1n0us** [comment]
good to know. thanks

________________________________

File: test/per_file_coverage.sh
Comment (leyao-daily) [reply test/per_file_coverage.sh:59]: Got it. So it seems the coverage-check only check nullvm + v8 under wasm_runtime?

Comment (PiotrSikora) [reply test/per_file_coverage.sh]: Yes, coverage is tested only with default build options (V8 and NullVM).

**Comment by PiotrSikora** [comment]
> it based on “llvm 10 config” and plus manually adjustment. literally “homemade”.

That's not great, but hopefully it won't be an issue for the maintainers.

**Comment by lum1n0us** [comment]
let me see what we can do

________________________________

File: bazel/envoy_build_system.bzl
Comment (PiotrSikora) [inline bazel/envoy_build_system.bzl:214]: Nit: move up to sort alphabetically.

File: bazel/foreign_cc/BUILD
Comment (PiotrSikora) [inline bazel/foreign_cc/BUILD:353]: Nit: move up to sort alphabetically.

File: bazel/repositories.bzl
Comment (PiotrSikora) [inline bazel/repositories.bzl:943]: Nit: move up to sort alphabetically.

File: bazel/repository_locations.bzl
Comment (PiotrSikora) [inline bazel/repository_locations.bzl:969]: Please update to the merged commit and make sure to update release date in UTC time (2021-05-07).

File: source/extensions/wasm_runtime/wamr/BUILD
Comment (PiotrSikora) [reply source/extensions/wasm_runtime/wamr/BUILD:16]: This is still needed.

**Review by PiotrSikora — DISMISSED** [review]
A few small changes (style & commit update), but LGTM otherwise. Thanks a lot!

File: bazel/envoy_build_system.bzl
Comment (leyao-daily) [reply bazel/envoy_build_system.bzl:214]: OK, Should I move `envoy_select_wasm_v8 = _envoy_select_wasm_v8` up to sort alphabetically together? LOL

Comment (PiotrSikora) [reply bazel/envoy_build_system.bzl:214]: If you could, thanks and sorry for nitpicking!

I'm actually surprised that format check doesn't complain about this, I believe it used to in the past.

Comment (leyao-daily) [reply bazel/envoy_build_system.bzl]: LOL. Fixed.

**Review by PiotrSikora — APPROVED** [review]
Awesome, thank you!

**Review by PiotrSikora — APPROVED** [review]

**Review by PiotrSikora — APPROVED** [review]
Thanks!

**Review by htuch — APPROVED** [review]
/lgtm deps

**Comment by htuch** [comment]
Thanks @leyao-daily. Needs main merge.

**Review by PiotrSikora — DISMISSED** [review]
@leyao-daily in the future, please use `git merge origin/main` instead of rebase, since it makes reviewing incremenal changes harder to follow, and GitHub is dismissing previous approvals after each sync.

**Comment by leyao-daily** [comment]
> @leyao-daily in the future, please use `git merge origin/main` instead of rebase, since it makes reviewing incremenal changes harder to follow, and GitHub is dismissing previous approvals after each sync.

Ok. Get it. Thanks for your advice.

**Comment by PiotrSikora** [comment]
@moderation @lizan PTAL.

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:705]: WAMR appears to have regular releases - https://github.com/bytecodealliance/wasm-micro-runtime/releases/tag/WAMR-04-15-2021. Can we move to a release vs. commit?

**Review by moderation — COMMENTED** [review]
Can we use WAMR release vs. commit?

File: bazel/repository_locations.bzl
Comment (leyao-daily) [reply bazel/repository_locations.bzl:705]: The PR to support Envoy in WAMR hasn't been released. We use commit to utilize the latest updates.

Comment (lum1n0us) [reply bazel/repository_locations.bzl:705]: Since WAMR team still has several ongoing developments about *Envoy integration*. We'd like to keep using commits for agility.

Comment (moderation) [reply bazel/repository_locations.bzl:705]: As per the dependency policy, releases are preferred over commits (particularly when there are regular releases) - https://github.com/envoyproxy/envoy/blob/main/DEPENDENCY_POLICY.md#declaring-external-dependencies. Once the WAMR Envoy integration stabilizes the plan should be to move to releases

Comment (leyao-daily) [reply bazel/repository_locations.bzl:705]: Ok. Got it. Thanks. We will move to release when WAMR releases in the future.

**Comment by leyao-daily** [comment]
About the API change, what should I do to pass the check?

**Comment by PiotrSikora** [comment]
> About the API change, what should I do to pass the check?

Wait for one of the @envoyproxy/api-shepherds to approve.

**Comment by PiotrSikora** [comment]
@leyao-daily this needs merge with master to resolve conflicts.

@htuch @lizan could one of you please take a look at this PR?

**Comment by leyao-daily** [comment]
Once the CI/CD pass, I think this PR is ready to be merged.

File: source/extensions/extensions_metadata.yaml
Comment (htuch) [inline source/extensions/extensions_metadata.yaml:671]: @leyao-daily can you put move thte build comments I previously asked for to YAML comments here? Thanks.
/wait

Comment (leyao-daily) [reply source/extensions/extensions_metadata.yaml:671]: Got it. And same for wasmtime and wavm?

**Review by htuch — APPROVED** [review]
/lgtm deps

**Pull request merged**
